File: app/src/main/java/com/nuhkoca/myapplication/ui/video/VideoViewModel.java
package com.nuhkoca.myapplication.ui.video;

import com.nuhkoca.myapplication.data.remote.player.PlayerResponse;

import javax.inject.Inject;

import androidx.annotation.NonNull;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;
import io.reactivex.observers.DisposableSingleObserver;

/**
 * A {@link androidx.lifecycle.ViewModel} for media playing
 *
 * @author nuhkoca
 */
public class VideoViewModel extends ViewModel {

    private CompositeDisposable compositeDisposable;
    private PlayerUseCase playerUseCase;

    private MutableLiveData<PlayerResponse> mPlayerResponse = new MutableLiveData<>();

    /**
     * A default constructor that inject required dependencies
     *
     * @param compositeDisposable represents an instance of {@link CompositeDisposable}
     * @param playerUseCase       represents an instance of {@link PlayerUseCase}
     */
    @Inject
    VideoViewModel(@NonNull CompositeDisposable compositeDisposable,
                   @NonNull PlayerUseCase playerUseCase) {
        this.compositeDisposable = compositeDisposable;
        this.playerUseCase = playerUseCase;
    }

    /**
     * Returns playable content
     *
     * @param videoId indicates the video id
     */
    void getPlayableContent(@NonNull String videoId) {
        Disposable url = playerUseCase.execute(new DisposableSingleObserver<PlayerResponse>() {
            @Override
            public void onSuccess(PlayerResponse playerResponse) {
                mPlayerResponse.setValue(playerResponse);
            }

            @Override
            public void onError(Throwable e) {

            }
        }, videoId);

        compositeDisposable.add(url);
    }

    /**
     * Returns the playable content
     *
     * @return the playable content
     */
    MutableLiveData<PlayerResponse> getContent() {
        return mPlayerResponse;
    }

    /**
     * Clears references
     */
    @Override
    protected void onCleared() {
        compositeDisposable.clear();
        super.onCleared();
    }
}

File: app/src/main/java/com/nuhkoca/myapplication/ui/video/VideoViewModel.kt
package com.nuhkoca.myapplication.ui.video

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import com.nuhkoca.myapplication.data.remote.player.PlayerResponse
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.observers.DisposableSingleObserver
import javax.inject.Inject

/**
 * A [androidx.lifecycle.ViewModel] for media playing
 *
 * @param compositeDisposable represents an instance of [CompositeDisposable]
 * @param playerUseCase       represents an instance of [PlayerUseCase]
 *
 * @author nuhkoca
 */
class VideoViewModel @Inject constructor(
    private val compositeDisposable: CompositeDisposable,
    private val playerUseCase: PlayerUseCase
) : ViewModel() {
    /**
     * Returns the playable content
     *
     * @return the playable content
     */
    private val _content = MutableLiveData<PlayerResponse>()
    val content: LiveData<PlayerResponse> get() = _content

    /**
     * Returns playable content
     *
     * @param videoId indicates the video id
     */
    fun getPlayableContent(videoId: String) {
        val url = playerUseCase.execute(object : DisposableSingleObserver<PlayerResponse>() {
            override fun onSuccess(playerResponse: PlayerResponse) {
                _content.value = playerResponse
            }

            override fun onError(e: Throwable) {

            }
        }, videoId)

        compositeDisposable.add(url!!)
    }

    /**
     * Clears references
     */
    override fun onCleared() {
        compositeDisposable.clear()
        super.onCleared()
    }
}
